engine/src/main/java/io/zeebe/engine/processing/streamprocessor/writers/TypedStreamWriterImpl.java

```diff
@@ -7,20 +7,114 @@
  */
 package io.zeebe.engine.processing.streamprocessor.writers;
 
+import static io.zeebe.engine.processing.streamprocessor.TypedEventRegistry.EVENT_REGISTRY;
+
 import io.zeebe.engine.processing.streamprocessor.TypedRecord;
 import io.zeebe.logstreams.log.LogStreamBatchWriter;
+import io.zeebe.logstreams.log.LogStreamBatchWriter.LogEntryBuilder;
 import io.zeebe.msgpack.UnpackedObject;
 import io.zeebe.protocol.impl.record.RecordMetadata;
 import io.zeebe.protocol.record.RecordType;
 import io.zeebe.protocol.record.RejectionType;
+import io.zeebe.protocol.record.ValueType;
 import io.zeebe.protocol.record.intent.Intent;
-import java.util.function.Consumer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.UnaryOperator;
+
+public class TypedStreamWriterImpl implements TypedStreamWriter {
+
+  private static final UnaryOperator<RecordMetadata> NO_MODIFIER = UnaryOperator.identity();
 
-public final class TypedStreamWriterImpl extends TypedCommandWriterImpl
-    implements TypedStreamWriter {
+  private final Map<Class<? extends UnpackedObject>, ValueType> typeRegistry;
+  private final RecordMetadata metadata = new RecordMetadata();
+  private final LogStreamBatchWriter batchWriter;
+
+  private long sourceRecordPosition = -1;
 
   public TypedStreamWriterImpl(final LogStreamBatchWriter batchWriter) {
-    super(batchWriter);
+    this.batchWriter = batchWriter;
+    typeRegistry = new HashMap<>();
+    EVENT_REGISTRY.forEach((e, c) -> typeRegistry.put(c, e));
+  }
+
+  protected void initMetadata(
+      final RecordType type, final Intent intent, final UnpackedObject value) {
+    metadata.reset();
+    final ValueType valueType = typeRegistry.get(value.getClass());
+    if (valueType == null) {
+      // usually happens when the record is not registered at the TypedStreamEnvironment
+      throw new RuntimeException("Missing value type mapping for record: " + value.getClass());
+    }
+
+    metadata.recordType(type).valueType(valueType).intent(intent);
+  }
+
+  protected void appendRecord(
+      final long key,
+      final RecordType type,
+      final Intent intent,
+      final UnpackedObject value,
+      final UnaryOperator<RecordMetadata> modifier) {
+    appendRecord(key, type, intent, RejectionType.NULL_VAL, "", value, modifier);
+  }
+
+  protected void appendRecord(
+      final long key,
+      final RecordType type,
+      final Intent intent,
+      final RejectionType rejectionType,
+      final String rejectionReason,
+      final UnpackedObject value,
+      final UnaryOperator<RecordMetadata> modifier) {
```

Comment:
The modifier is no longer called.

Reply:
It is, but just a couple line lower on L87. 

You can provide a completely new metadata object in the modifier, so we need to use the return value now.

Reply:
Sorry, my bad